### PR TITLE
[WIP] Find unused ports more reliably

### PR DESCRIPTION
**Pull request by gpang**
Passing ServerSocket port 0 sometimes returns a socket that has a conflicting port (by another process). However, explicitly passing in a port number does not have the same issue. Therefore, we must manually find a random port.

Hopefully, this improves the flaky integration tests.

### CODE REVIEW

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
This test seems to have a different reason for flakiness.
```
alluxio.server.ft.journal.ufs.UfsJournalIntegrationTest.directory(alluxio.server.ft.journal.ufs.UfsJournalIntegrationTest)
  Run 1: UfsJournalIntegrationTest.directory:424->directoryTestUtil:428->createFsMasterFromJournal:635 » Runtime
  Run 2: UfsJournalIntegrationTest.directory:424->directoryTestUtil:428->createFsMasterFromJournal:635 » Runtime
```

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
another port-unrelated flaky test:
```
Tests run: 5, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 39.345 sec <<< FAILURE! - in alluxio.server.ft.MultiWorkerIntegrationTest
writeLargeFile(alluxio.server.ft.MultiWorkerIntegrationTest)  Time elapsed: 7.423 sec  <<< FAILURE!
java.lang.AssertionError: expected:<100> but was:<50>
	at alluxio.server.ft.MultiWorkerIntegrationTest.writeLargeFile(MultiWorkerIntegrationTest.java:100)
```

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by ZacBlanco** [comment]
@gpang it looks like the tests are still flaky with this PR. From what I see there might still be some failures due to port conflicts

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by ZacBlanco** [comment]
jenkins, test this please

File: tests/src/test/java/alluxio/master/PortRegistryTest.java [hunk 1]
```diff
@@ -86,4 +94,54 @@ public void releaseDeletesFile() {
     }
     assertThat(successes, greaterThan(2));
   }
+
+  @Test
+  public void reserveSamePort() throws Exception {
```
Comment (chcn) [inline tests/src/test/java/alluxio/master/PortRegistryTest.java]: what's the purpose of this test?

**Comment by yuzhu** [comment]
jenkins, test this please

**Comment by gpang** [comment]
This will not fix the problems yet. I think I know what the solution is, but this current PR will not completely fix it.

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by ZacBlanco** [comment]
jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
alluxio-bot, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
Jenkins, test this please

**Pull request closed without merging**